### PR TITLE
[ray_client]: Add more retry logic

### LINKED ISSUE #13446: Ray client connect retries not sufficient: sometimes get 502 error
I don't believe the `channel_ready_future()` is sufficient to guarantee the connection is ready, since sometimes I still see a 502 error from the data client.

How about we issue a health check RPC and destroy/recreate the channel if that fails?

cc @barakmich 

```
Got Error from data channel -- shutting down: <_MultiThreadedRendezvous of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "Received http2 header with status: 502"
	debug_error_string = "{"created":"@1610606646.739755640","description":"Received http2 :status header with non-200 OK status","file":"src/core/ext/filters/http/client/http_client_filter.cc","file_line":129,"grpc_message":"Received http2 header with status: 502","grpc_status":14,"value":"502"}"
>
Got Error from logger channel -- shutting down: <_MultiThreadedRendezvous of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "Received http2 header with status: 502"
	debug_error_string = "{"created":"@1610606646.741104491","description":"Received http2 :status header with non-200 OK status","file":"src/core/ext/filters/http/client/http_client_filter.cc","file_line":129,"grpc_message":"Received http2 header with status: 502","grpc_status":14,"value":"502"}"
>
Exception in thread Thread-5:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/eric/Desktop/ray/python/ray/util/client/dataclient.py", line 76, in _data_main
    raise e
  File "/home/eric/Desktop/ray/python/ray/util/client/dataclient.py", line 61, in _data_main
    for response in resp_stream:
  File "/home/eric/.local/lib/python3.6/site-packages/grpc/_channel.py", line 416, in __next__
    return self._next()
  File "/home/eric/.local/lib/python3.6/site-packages/grpc/_channel.py", line 706, in _next
    raise self
grpc._channel._MultiThreadedRendezvous: <_MultiThreadedRendezvous of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "Received http2 header with status: 502"
	debug_error_string = "{"created":"@1610606646.739755640","description":"Received http2 :status header with non-200 OK status","file":"src/core/ext/filters/http/client/http_client_filter.cc","file_line":129,"grpc_message":"Received http2 header with status: 502","grpc_status":14,"value":"502"}"
```

### PR DESCRIPTION
**Pull request by barakmich**
## Related issue number

Closes #13446 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/util/client/worker.py [hunk 1]
```diff
@@ -68,14 +78,58 @@ def __init__(self,
...
+        while not ray_ready:
+            if conn_attempts >= (connection_retries + 1):
+                raise ConnectionError(
+                    "ray client connection timeout: ray not initialized")
+            logger.info("Waiting for Ray to become ready on the server, "
+                        f", retry in {timeout}s...")
+            time.sleep(timeout)
+            timeout = backoff(timeout)
+            conn_attempts += 1
+            ray_ready = self.is_initialized()
```
Comment (ericl) [inline python/ray/util/client/worker.py:130]: Do we need three different retry loops for the different stages? Why not have just a single one that tests this last condition?

**Review by ericl — CHANGES_REQUESTED** [review]
Comment on simplifying

**Comment by barakmich** [comment]
Because it could fail at any one of these stages.

Now, one big retry loop would look like a finite state machine of ```
while retrying:
   if in step A:
   if in step B:
   if in step C:
```

But we can't "test this last condition" alone. If this is what you'd prefer, I'll go do that.

**Comment by ericl** [comment]
Why can't we test the last condition alone? C implies {A, B} are fine right? I think a loop that just tests C would be sufficient. If it's for better error messages, we could also check {A, B} prior to C in the loop.

**Comment by ericl** [comment]
Btw, it would be good to have a test here

File: python/ray/util/client/worker.py [hunk 2]
```diff
@@ -59,23 +71,68 @@ def __init__(self,
...
             self.channel = grpc.insecure_channel(conn_str)
 
+        # Retry the connection until the channel responds to something
+        # looking like a gRPC connection, though it may be a proxy.
+        #
+        # This is a finite state machine following the progression of states
+        # a connecting client should check and retry for.
         conn_attempts = 0
         timeout = INITIAL_TIMEOUT_SEC
+        connection_state = CONN_STATE_CHANNEL
```
Comment (ericl) [inline python/ray/util/client/worker.py:81]: Why do we need a FSM? Why not reset the state to the initial state on connection failure and retry end-to-end?

If the server is up, the connection should be instant so there isn't any benefit in keeping intermediate states around on failure.

File: python/ray/tests/test_client.py [hunk 3]
```diff
@@ -373,5 +376,22 @@ def test_internal_kv(ray_start_regular_shared):
         assert ray._internal_kv_get("apple") == b""
 
 
+def test_startup_retry(ray_start_regular_shared):
+    from ray.util.client import ray as ray_client
+    with pytest.raises(ConnectionError):
+        ray_client.connect("localhost:50051", connection_retries=1)
+
+    def run_server():
+        server = ray_client_server.serve("localhost:50051")
```
Comment (ericl) [inline python/ray/tests/test_client.py:385]: Add a delay before the server starts?

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
```
//python/ray/tests:test_client                                          TIMEOUT in 3 out of 3 in 75.0s
  Stats over 3 runs: max = 75.0s, min = 75.0s, avg = 75.0s, dev = 0.0s
  /home/travis/.cache/bazel/_bazel_travis/b88c129a127452fc94033a29d9f90e20/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_client/test.log
  /home/travis/.cache/bazel/_bazel_travis/b88c129a127452fc94033a29d9f90e20/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_client/test_attempts/attempt_1.log
  /home/travis/.cache/bazel/_bazel_travis/b88c129a127452fc94033a29d9f90e20/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_client/test_attempts/attempt_2.log
```

**Comment by ericl** [comment]
Tests / lint still failing

**Comment by barakmich** [comment]
I'm starting to run out of ideas, but we'll see if this goes

**Comment by barakmich** [comment]
I think I figured it out

**Comment by barakmich** [comment]
This time for sure -- I got the reproduction locally and tracked it down

**Pull request merged**